### ISSUE TITLE
Renku file resource

### ISSUE THREAD

**Issue by jachro** [SwissDataScienceCenter/renku#696]
As a Renku owner, I'd like files to be uniquely identified by a queryable URL resource.

**Comment by rokroskar** [comment]
We need to discuss this a bit more - it's not obvious to me what the URL should be. Files belong to projects, so should it be something like `http(s)://renku/entities/projects/files/...`? Or based on your suggestion in SwissDataScienceCenter/renku-gateway/issues/179 files could have ids? But how do we generate the ids?

**Comment by jachro** [comment]
Yes, it's not a trivial one. What about files not being linked to any
project or dataset by direct links from a file to project or dataset? The
reverse links are more or less obvious, I mean project/dataset has a file.
The reason for that is, according to what we discussed last Thursday, files
not necessarily have to be part of a project so maybe they can be simple
pointers to some file in storage like S3, network drive or dataset storage
(Zenodo).

In terms of the ids, what if they can be the hash/checksum/or whatever
better deterministic representation of a file? That would have to be the
hash/checksum/whatever of the file on adding to the project/dataset maybe?

**Comment by rokroskar** [comment]
> What about files not being linked to any project or dataset by direct links from a file to project or dataset

you mean that a file has some unique identifier and the datasets/projects just link to that identifier? That would be an option, the question we loop back to is just what that identifier should be.

**Comment by jachro** [comment]
True. What if the identifier is simply something we derive from file own
metadata like name, size, date of creation, checksum, etc? I believe having
some deterministic way of calculating it could be really beneficial.

**Comment by rokroskar** [comment]
Relevant for this discussion: https://researchobject.github.io/ro-crate/0.2/

There are some useful ontologies mentioned there for describing files and repositories. We should consider reworking the metadata as a part of this epic.

**Comment by rokroskar** [comment]
More things to consider: [DATS](https://www.nature.com/articles/sdata201759) and [data publishing best practices](https://www.w3.org/TR/dwbp/)

**Comment by rokroskar** [comment]
A good collection of resources about technical metadata: https://wiki.duraspace.org/display/samvera/Technical+Metadata+Application+Profile

**Comment by rokroskar** [comment]
A possible graph to keep in mind: 

![Untitled](https://user-images.githubusercontent.com/3353586/68717569-6cb99380-05a7-11ea-8d2b-cb9c9cdcee61.png)

**Comment by rokroskar** [comment]
The API endpoints would look something like:

### Files

```
https://renku/entities/files
```

Returns a list of `File` objects and the blobs associated with each one. 

#### Optional query parameters

* `project`
* `dataset`
* `path`

#### URI 

`https://renku/entities/projects/<project>/files/<path>` 

It translates to a query of `https://renku/entities/files?project=<project>&path=<path>`. `project` and `path` both need to be URL-encoded. 


### Blobs

```
https://renku/entities/blobs/<blob-id>
```

Returns a list of `Blob` objects.

Optional query parameters:

* `file URI`
* `project`
* `dataset`

If `<blob-id>` is provided, the detailed metadata of the blob is returned. `blob-id` should be a hash and we could leave the choice of hash function open by using the conventrion like e.g. `sha256:<hash>` (see also [naming things with hashes](https://tools.ietf.org/html/rfc6920)).

### Changes to existing API

We already have the `projects` and `datasets` endpoints so it would make sense to try to implement the files and blobs (which are currently missing) in a way that does not require a rewrite of the entire system. This will lead to some potential inconsistencies while the API stabilizes, but should allow us to move forward. For example, if the primary way to begin interacting with the API is via the `https://renku/entities/projects/<project>/files/<path>` endpoint, then we can hide the inconsistency for the time being.

**Comment by lorenzo-cavazzi** [comment]
I would add `commit` to the picture. I believe it's pretty common to query `files`/`blobs` per commit, especially from the UI.

Since `files` are the abstraction of the `blobs`, I am wondering if it also makes sense to add `path` directly in the blobs' query parameters instead of allowing only the use of `file URI`.

**Comment by rokroskar** [comment]
You mean commit as an entity or as a property?

**Comment by lorenzo-cavazzi** [comment]
I am not sure which solution is better. Since I imagine using commits as a filter, I was initially thinking about adding them as a property in the blobs, but if our final goal is having all the knowledge in the KG it may be useful to treat the commits as first-class citizens and add them as an entity.

In this way, we will be able to query commits from an API similar to the ones proposed above with a set of optional query parameters.

```
https://renku/entities/commits
```

**Comment by jachro** [comment]
There's also the issue we didn't solve last time about files and blobs linked to projects directly like README.md. As far as I remember we were not sure if such file entity should be shared between different projects or not. If a file entity will simply contain some very common properties like name it could be a bit hard to create different file entities for different projects. From the other side if a file entity is shared but points to different blobs which then are linked to different projects, maybe we are fine?

**Comment by rokroskar** [comment]
I don't think it should be shared - its URI should be e.g. `https://renku/entities/projects/<project>/files/README.md` which by definition will not be shared between different projects.

**Comment by jachro** [comment]
However, one question comes to my mind. Should it be `https://renku/entities/projects/<project>/files/README.md` or `https://renku/entities/files?project=<project>&file=README.md`?

**Comment by rokroskar** [comment]
See the `URI` part above - I think we should offer the full URL as a convenience and redirect to `https://renku/entities/files?project=<project>&file=README.md`.

**Comment by jachro** [comment]
Right. The full URL I'm happy with but in terms of the files resource the `https://renku/entities/files/README.md?project=<project>` resonates a bit more to me as I understand the file entity name would be `README.md` and you filter the results for a concrete project. What do you think?

**Comment by rokroskar** [comment]
Basically what you are suggesting is something like this:

![Untitled](https://user-images.githubusercontent.com/3353586/69164209-d974e680-0aef-11ea-8ee6-53b9cc5b03c3.png)

**Comment by jachro** [comment]
Yes, that's what I was thinking about but to be honest, although right from a logical point of view,  I'm wondering whether that modelling makes much of sense in the projects/files context. The questions I'm asking myself are:
* Does sharing files metadata between different project make sense? Are README files having the same name but in different projects the same? At first glance, it looks to me that they are simply by some reason just having the same name but in fact, they are different beings. From the other side if the File entity describes some other file properties like e.g. `purpose`, then sharing single README File entity with the `purpose` of `A file describing a project`, makes much more sense.
* Is situation the same in the case of data-set files? That looks much straightforward to me as if you have an imported Zenodo dataset then you definitely share the same File entities.

**Comment by rokroskar** [comment]
Right, I also don't feel comfortable sharing entities between projects based only on the filename. That's why I suggested that the URIs are actually different (e.g. include the project name and namespace). But I don't know how best to reconcile that with the need for an `entities/files` endpoint.

**Comment by rokroskar** [comment]
Now that I think about it, we are effectively already doing that - we have triples like 

```
<file-id> `atLocation` <path>
```

and `<path>` is just a string, so several entities could point to the same `<path>`. I guess there is nothing wrong with that.

**Comment by jachro** [comment]
Right, so does this mean that we'd have a single file, let's say README, with several `atLocation`?

**Comment by rokroskar** [comment]
no I mean `<path>` is relative to the project root, so several `<file-id>` might share the same `<path>`.

**Comment by rokroskar** [comment]
I've made a [notebook](https://renkulab.io/projects/rok.roskar/ontology-playground/files/blob/notebooks/file-entity.ipynb) to try and make some of the aspects of this question more concrete. Feel free to fork the project and make changes. 

The basic realization boils down to: the KG should facilitate discovery of links across projects and other entities (e.g. datasets) by virtue of data. We should include enough information in the KG for establishing provenance, but for detailed _repository-specific_ versioning, we should always resort to git and not try to replicate the entire git graph and logic in the KG. So I shouldn't try to ask the KG to give me all the files in the tree of a repository at a given commit - but I can ask the KG to give me all versions of a file and all of the locations of those versions across projects.

**Comment by lorenzo-cavazzi** [comment]
The notebook is a great idea!

I tried to understand this logic and played a bit with it. After some initial concerns about spamming `schema:Path`, I think this approach is not bad.

Anyway, I think dataset should be elaborated a bit more.
How can I know from that graph that `abcd-1234` and `abcd-5678` are 2 different versions of the dataset `myDataset`?
Or, in other words, how could this graph answer to the following query
```
https://<renku>/entities/blobs?project=https://renkulab.io/projects/1&dataset=myDataset
```

**Comment by rokroskar** [comment]
Ah yes, I didn't quite finish the datasets. I think it would be the same relationship as with projects. 

I had the same concern about too many `path` triples, but I think if we stick to the idea of only putting the equivalent of git diffs into the graph and not the entire git tree this can work. There is probably some query that can give you the answer to "which files are present in the project at revision `<ref>`?" but I don't think that should be our primary concern/use of the KG.

**Comment by rokroskar** [comment]
I added the `mydataset` which has incoming edges of `isVersionOf` from the dataset versions. I also realized that we want the `isPartOf` edges to be used on the same level of property - i.e. if `blob --atLocation--> location --isPartOf--> ref --isVersionOf--> project` then it's nice to have a similar relationship for the datasets. Which makes me think that we need `location` also for "within a dataset". I.e. I can have this project structure:

```
project/ 
  data/
    mydataset/
      file.csv
```
  
Then the dataset entity is really separate from the project, although it can be a part of the project. The queries then reflect this, so if you look at the last query in the notebook now which poses the question "given this file at a particular location in this project, tell me where else it exists" it returns

```
(rdflib.term.URIRef('https://renkulab.io/datasets/mydataset/1234'), rdflib.term.Literal('file.csv'))
(rdflib.term.URIRef('https://renkulab.io/projects/2/commit/aeio'), rdflib.term.Literal('file.csv'))
(rdflib.term.URIRef('https://renkulab.io/projects/1/commit/abcd'), rdflib.term.Literal('data/mydataset/file.csv'))
```

So datasets and projects are actually accessed with the same path query - of course you can always filter on type if you want one or the other (I haven't added types in the notebook because it clutters the graph).

**Comment by rokroskar** [comment]
updated graph:

![image](https://user-images.githubusercontent.com/3353586/74536115-b4014880-4f37-11ea-8425-89a586420834.png)

**Comment by jachro** [comment]
I have to say I quite like it. The only thing which wasn't entirely natural for me was a commit that `isVersionOf` a project. After some thinking, I realized it makes sense to do it this way.
👍

**Comment by rokroskar** [comment]
what seems unnatural? how would you change it?

**Comment by jachro** [comment]
I think it's ok but at a first glance, it wasn't somehow obvious to me that commit is a version of a project. But actually it is and it makes a lot of sense to model it that way. Sorry, I should even not make a comment about it as it raises some unnecessary confusion.

**Comment by rokroskar** [comment]
Haha no worries I think it's good to get all the confusion out of the way early!
